File: php/mercadobitcoin_sdk.php
<?php
declare(strict_types=1);

// MercadoBitcoin SDK

require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/UtilityType.php';
require_once __DIR__ . '/core/Spec.php';
require_once __DIR__ . '/core/Helpers.php';

// Load utility registration
require_once __DIR__ . '/utility/Register.php';

// Load config and features
require_once __DIR__ . '/config.php';
require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/features.php';

use Voxgig\Struct\Struct;

class MercadoBitcoinSDK
{
    public string $mode;
    public array $features;
    public ?array $options;

    private $_utility;
    private $_rootctx;

    public function __construct(array $options = [])
    {
        $this->mode = "live";
        $this->features = [];
        $this->options = null;

        $utility = new MercadoBitcoinUtility();
        $this->_utility = $utility;

        $config = MercadoBitcoinConfig::make_config();

        $this->_rootctx = ($utility->make_context)([
            "client" => $this,
            "utility" => $utility,
            "config" => $config,
            "options" => $options ?? [],
            "shared" => [],
        ], null);

        $this->options = ($utility->make_options)($this->_rootctx);

        if (Struct::getpath($this->options, "feature.test.active") === true) {
            $this->mode = "test";
        }

        $this->_rootctx->options = $this->options;

        // Add features from config.
        $feature_opts = MercadoBitcoinHelpers::to_map(Struct::getprop($this->options, "feature"));
        if ($feature_opts) {
            $items = Struct::items($feature_opts);
            if ($items) {
                foreach ($items as $item) {
                    $fname = $item[0];
                    $fopts = MercadoBitcoinHelpers::to_map($item[1]);
                    if ($fopts && isset($fopts["active"]) && $fopts["active"] === true) {
                        ($utility->feature_add)($this->_rootctx, MercadoBitcoinFeatures::make_feature($fname));
                    }
                }
            }
        }

        // Add extension features.
        $extend_val = Struct::getprop($this->options, "extend");
        if (is_array($extend_val)) {
            foreach ($extend_val as $f) {
                if (is_object($f) && method_exists($f, 'get_name')) {
                    ($utility->feature_add)($this->_rootctx, $f);
                }
            }
        }

        // Initialize features.
        foreach ($this->features as $f) {
            ($utility->feature_init)($this->_rootctx, $f);
        }

        ($utility->feature_hook)($this->_rootctx, "PostConstruct");
    }

    public function options_map(): array
    {
        $out = Struct::clone($this->options);
        return is_array($out) ? $out : [];
    }

    public function get_utility()
    {
        return MercadoBitcoinUtility::copy($this->_utility);
    }

    public function get_root_ctx()
    {
        return $this->_rootctx;
    }

    public function prepare(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;
        $fetchargs = $fetchargs ?? [];

        $ctrl = MercadoBitcoinHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "prepare",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $opts = $this->options;
        $path = Struct::getprop($fetchargs, "path") ?? "";
        $path = is_string($path) ? $path : "";
        $method_val = Struct::getprop($fetchargs, "method") ?? "GET";
        $method_val = is_string($method_val) ? $method_val : "GET";
        $params = MercadoBitcoinHelpers::to_map(Struct::getprop($fetchargs, "params")) ?? [];
        $query = MercadoBitcoinHelpers::to_map(Struct::getprop($fetchargs, "query")) ?? [];
        $headers = ($utility->prepare_headers)($ctx);

        $base = Struct::getprop($opts, "base") ?? "";
        $base = is_string($base) ? $base : "";
        $prefix = Struct::getprop($opts, "prefix") ?? "";
        $prefix = is_string($prefix) ? $prefix : "";
        $suffix = Struct::getprop($opts, "suffix") ?? "";
        $suffix = is_string($suffix) ? $suffix : "";

        $ctx->spec = new MercadoBitcoinSpec([
            "base" => $base, "prefix" => $prefix, "suffix" => $suffix,
            "path" => $path, "method" => $method_val,
            "params" => $params, "query" => $query, "headers" => $headers,
            "body" => Struct::getprop($fetchargs, "body"),
            "step" => "start",
        ]);

        // Merge user-provided headers.
        $uh = Struct::getprop($fetchargs, "headers");
        if (is_array($uh)) {
            foreach ($uh as $k => $v) {
                $ctx->spec->headers[$k] = $v;
            }
        }

        [$_, $err] = ($utility->prepare_auth)($ctx);
        if ($err) {
            return ($utility->make_error)($ctx, $err);
        }

        [$fetchdef, $fd_err] = ($utility->make_fetch_def)($ctx);
        if ($fd_err) {
            return ($utility->make_error)($ctx, $fd_err);
        }
        return $fetchdef;
    }

    public function direct(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;

        // direct() is the raw-HTTP escape hatch: it never throws, it returns
        // an {ok, err, ...} dict. prepare() now raises on error, so catch it
        // and surface the failure through the dict instead.
        try {
            $fetchdef = $this->prepare($fetchargs);
        } catch (\Throwable $err) {
            return ["ok" => false, "err" => $err];
        }

        $fetchargs = $fetchargs ?? [];
        $ctrl = MercadoBitcoinHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "direct",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $url = $fetchdef["url"] ?? "";
        [$fetched, $fetch_err] = ($utility->fetcher)($ctx, $url, $fetchdef);

        if ($fetch_err) {
            return ["ok" => false, "err" => $fetch_err];
        }

        if ($fetched === null) {
            return [
                "ok" => false,
                "err" => $ctx->make_error("direct_no_response", "response: undefined"),
            ];
        }

        if (is_array($fetched)) {
            $status = MercadoBitcoinHelpers::to_int(Struct::getprop($fetched, "status"));
            $headers = Struct::getprop($fetched, "headers") ?? [];

            // No-body responses (204, 304) and explicit zero content-length
            // must skip JSON parsing — calling json() on an empty body errors.
            $content_length = is_array($headers) ? ($headers["content-length"] ?? null) : null;
            $no_body = $status === 204 || $status === 304 || (string)$content_length === "0";

            $json_data = null;
            if (!$no_body) {
                $jf = Struct::getprop($fetched, "json");
                if (is_callable($jf)) {
                    try {
                        $json_data = $jf();
                    } catch (\Throwable $e) {
                        // Non-JSON body — leave data null but keep status/ok.
                        $json_data = null;
                    }
                }
            }

            return [
                "ok" => $status >= 200 && $status < 300,
                "status" => $status,
                "headers" => Struct::getprop($fetched, "headers"),
                "data" => $json_data,
            ];
        }

        return [
            "ok" => false,
            "err" => $ctx->make_error("direct_invalid", "invalid response type"),
        ];
    }


    private $_balance = null;

    // Idiomatic facade: $client->balance()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Balance() (PHP method
    // names are case-insensitive).
    public function balance($data = null)
    {
        require_once __DIR__ . '/entity/balance_entity.php';
        if ($data === null) {
            if ($this->_balance === null) {
                $this->_balance = new BalanceEntity($this, null);
            }
            return $this->_balance;
        }
        return new BalanceEntity($this, $data);
    }


    private $_candle = null;

    // Idiomatic facade: $client->candle()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Candle() (PHP method
    // names are case-insensitive).
    public function candle($data = null)
    {
        require_once __DIR__ . '/entity/candle_entity.php';
        if ($data === null) {
            if ($this->_candle === null) {
                $this->_candle = new CandleEntity($this, null);
            }
            return $this->_candle;
        }
        return new CandleEntity($this, $data);
    }


    private $_deposit_address = null;

    // Idiomatic facade: $client->deposit_address()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias DepositAddress() (PHP method
    // names are case-insensitive).
    public function deposit_address($data = null)
    {
        require_once __DIR__ . '/entity/deposit_address_entity.php';
        if ($data === null) {
            if ($this->_deposit_address === null) {
                $this->_deposit_address = new DepositAddressEntity($this, null);
            }
            return $this->_deposit_address;
        }
        return new DepositAddressEntity($this, $data);
    }


    private $_order = null;

    // Idiomatic facade: $client->order()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Order() (PHP method
    // names are case-insensitive).
    public function order($data = null)
    {
        require_once __DIR__ . '/entity/order_entity.php';
        if ($data === null) {
            if ($this->_order === null) {
                $this->_order = new OrderEntity($this, null);
            }
            return $this->_order;
        }
        return new OrderEntity($this, $data);
    }


    private $_order_book = null;

    // Idiomatic facade: $client->order_book()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias OrderBook() (PHP method
    // names are case-insensitive).
    public function order_book($data = null)
    {
        require_once __DIR__ . '/entity/order_book_entity.php';
        if ($data === null) {
            if ($this->_order_book === null) {
                $this->_order_book = new OrderBookEntity($this, null);
            }
            return $this->_order_book;
        }
        return new OrderBookEntity($this, $data);
    }


    private $_ticker = null;

    // Idiomatic facade: $client->ticker()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Ticker() (PHP method
    // names are case-insensitive).
    public function ticker($data = null)
    {
        require_once __DIR__ . '/entity/ticker_entity.php';
        if ($data === null) {
            if ($this->_ticker === null) {
                $this->_ticker = new TickerEntity($this, null);
            }
            return $this->_ticker;
        }
        return new TickerEntity($this, $data);
    }


    private $_trade = null;

    // Idiomatic facade: $client->trade()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Trade() (PHP method
    // names are case-insensitive).
    public function trade($data = null)
    {
        require_once __DIR__ . '/entity/trade_entity.php';
        if ($data === null) {
            if ($this->_trade === null) {
                $this->_trade = new TradeEntity($this, null);
            }
            return $this->_trade;
        }
        return new TradeEntity($this, $data);
    }


    private $_withdrawal = null;

    // Idiomatic facade: $client->withdrawal()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Withdrawal() (PHP method
    // names are case-insensitive).
    public function withdrawal($data = null)
    {
        require_once __DIR__ . '/entity/withdrawal_entity.php';
        if ($data === null) {
            if ($this->_withdrawal === null) {
                $this->_withdrawal = new WithdrawalEntity($this, null);
            }
            return $this->_withdrawal;
        }
        return new WithdrawalEntity($this, $data);
    }



    public static function test(?array $testopts = null, ?array $sdkopts = null): self
    {
        $sdkopts = $sdkopts ?? [];
        $sdkopts = Struct::clone($sdkopts);
        $sdkopts = is_array($sdkopts) ? $sdkopts : [];

        $testopts = $testopts ?? [];
        $testopts = Struct::clone($testopts);
        $testopts = is_array($testopts) ? $testopts : [];
        $testopts["active"] = true;

        if (!isset($sdkopts["feature"])) {
            $sdkopts["feature"] = [];
        }
        $sdkopts["feature"]["test"] = $testopts;

        $sdk = new MercadoBitcoinSDK($sdkopts);
        $sdk->mode = "test";
        return $sdk;
    }
}

File: lua/mercado-bitcoin_sdk.lua
-- MercadoBitcoin SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local MercadoBitcoinSDK = {}
MercadoBitcoinSDK.__index = MercadoBitcoinSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

MercadoBitcoinSDK._make_feature = _make_feature


function MercadoBitcoinSDK.new(options)
  local self = setmetatable({}, MercadoBitcoinSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features in the resolved order (make_options puts an explicit list
  -- order first, else defaults to test-first). Ordering matters: the `test`
  -- feature installs the base mock transport and the transport features
  -- (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
  -- must be added before them to sit at the base of the chain.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local featureorder = vs.getpath(self.options, "__derived__.featureorder")
    if type(featureorder) == "table" then
      for _, fname in ipairs(featureorder) do
        local fopts = helpers.to_map(feature_opts[fname])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function MercadoBitcoinSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function MercadoBitcoinSDK:get_utility()
  return Utility.copy(self._utility)
end


function MercadoBitcoinSDK:get_root_ctx()
  return self._rootctx
end


function MercadoBitcoinSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function MercadoBitcoinSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:Balance():list() / client:Balance():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MercadoBitcoinSDK:Balance(data)
  local EntityMod = require("entity.balance_entity")
  if data == nil then
    if self._balance == nil then
      self._balance = EntityMod.new(self, nil)
    end
    return self._balance
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Candle():list() / client:Candle():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MercadoBitcoinSDK:Candle(data)
  local EntityMod = require("entity.candle_entity")
  if data == nil then
    if self._candle == nil then
      self._candle = EntityMod.new(self, nil)
    end
    return self._candle
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:DepositAddress():list() / client:DepositAddress():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MercadoBitcoinSDK:DepositAddress(data)
  local EntityMod = require("entity.deposit_address_entity")
  if data == nil then
    if self._deposit_address == nil then
      self._deposit_address = EntityMod.new(self, nil)
    end
    return self._deposit_address
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Order():list() / client:Order():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MercadoBitcoinSDK:Order(data)
  local EntityMod = require("entity.order_entity")
  if data == nil then
    if self._order == nil then
      self._order = EntityMod.new(self, nil)
    end
    return self._order
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:OrderBook():list() / client:OrderBook():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MercadoBitcoinSDK:OrderBook(data)
  local EntityMod = require("entity.order_book_entity")
  if data == nil then
    if self._order_book == nil then
      self._order_book = EntityMod.new(self, nil)
    end
    return self._order_book
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Ticker():list() / client:Ticker():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MercadoBitcoinSDK:Ticker(data)
  local EntityMod = require("entity.ticker_entity")
  if data == nil then
    if self._ticker == nil then
      self._ticker = EntityMod.new(self, nil)
    end
    return self._ticker
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Trade():list() / client:Trade():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MercadoBitcoinSDK:Trade(data)
  local EntityMod = require("entity.trade_entity")
  if data == nil then
    if self._trade == nil then
      self._trade = EntityMod.new(self, nil)
    end
    return self._trade
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Withdrawal():list() / client:Withdrawal():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MercadoBitcoinSDK:Withdrawal(data)
  local EntityMod = require("entity.withdrawal_entity")
  if data == nil then
    if self._withdrawal == nil then
      self._withdrawal = EntityMod.new(self, nil)
    end
    return self._withdrawal
  end
  return EntityMod.new(self, data)
end




function MercadoBitcoinSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = MercadoBitcoinSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return MercadoBitcoinSDK
